### PR TITLE
Prevent removing flow that used by client flow overrides

### LINKED ISSUE #30707: prevent removing the flow when used by client flow overrides
See the parent issue https://github.com/keycloak/keycloak/issues/30706 for context.

### PR DESCRIPTION
**Pull request by lexcao**
Closes #30707

### CODE REVIEW

**Comment by ahus1** [comment]
@graziang - similar to the other PR, this is about a parent issue you've commented on. Maybe you can have a look. Thanks!

**Review by graziang — COMMENTED** [review]
Thanks @lexcao. I have a doubt, in the case of many clients, it might not be easy to identify the clients with a custom flow from the admin console. Should we instead force the removal of the custom flow from the clients when it is deleted? @mposolda WDYT?

**Comment by lexcao** [comment]
Hi @graziang 

> it might not be easy to identify the clients with a custom flow from the admin console.

We can have a quick overview on the flow for what clients specified on the new Admin UI.

image

File: server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java [hunk 1]
```diff
@@ -650,7 +647,7 @@ private static GroupModel findSubGroup(String[] segments, int index, GroupModel
                 } else {
                     if (index + 1 < pathSegments.length) {
                         GroupModel found = findSubGroup(pathSegments, index + 1, group);
-                        if (found != null) return found;
+                        return found;
```
Comment (mposolda) [inline server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java:650]: This change is not related to authentication flows. Did you added it by accident to this PR?

File: server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java [hunk 2]
```diff
@@ -922,6 +919,14 @@ public static boolean isFlowUsed(RealmModel realm, AuthenticationFlowModel model
         if ((realmFlow = realm.getDockerAuthenticationFlow()) != null && realmFlow.getId().equals(model.getId())) return true;
         if ((realmFlow = realm.getFirstBrokerLoginFlow()) != null && realmFlow.getId().equals(model.getId())) return true;
 
+        boolean overrideByClient = realm.getClientsStream()
```
Comment (mposolda) [inline server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java:922]: This is performance killer in case that realm has many clients.

There are those methods introduced https://github.com/keycloak/keycloak/pull/21058/files in some old PR. Could you please rather use those instead of `realms.getClientsStream()` with the programmatic filtering?

**Review by mposolda — CHANGES_REQUESTED** [review]
@lexcao @graziang Thanks for the PR and the review!

Added 2 minor inline comments. Could you please check them?

File: server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java (hunk 1 above)
Comment (mposolda) [reply server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java:650]: @lexcao Not sure why this comment was resolved? The PR was not updated.

Comment (lexcao) [reply server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java:650]: @mposolda 
Oh, sorry, I am fixing now, waiting for push the latest code.

**Comment by lexcao** [comment]
Hi @mposolda 
Thanks for the information, I just updated the code.

**Review by mposolda — APPROVED** [review]
@lexcao Thanks for the updates!
